### ISSUE TITLE
Double/duplicate payments since upgrading Prestashop to 1.6.1.5

### ISSUE THREAD

**Issue by luwes** [ICEPAY/Prestashop#1]
Any idea why we're getting these warnings since the upgrade? Thanks
(The actual money that comes in is just the one payment)

![screen shot 2016-05-11 at 10 12 11](https://cloud.githubusercontent.com/assets/360826/15174263/045427a2-1761-11e6-90f5-e92a7ac4a287.png)

**Comment by tsygankov** [comment]
Hi luwes,

Can you please specify which version of PrestaShop and ICEPAY Payment Module for Prestashop are you running?

**Comment by luwes** [comment]
Thanks for the reply. We were using Prestashop 1.6.0.9 first and the ICEPAY payment module is version 2.1.2.

**Comment by luwes** [comment]
The first payment seems to be from here, https://github.com/ICEPAYdev/Prestashop/blob/master/validate.php#L85
as there the invoice is added as well.

**Comment by tsygankov** [comment]
We've tested the module in Prestashop 1.6.0.9 and can't reproduce the problem. Looks like for some reason post-back notification was sent twice with the same status code. I will ask my colleges to check getaway logs.

Both payments are created in validate.php file that you mentioned above. For 2nd payment invoice is not created because setInvoice($use_existing_payment = false) method creates only one invoice per order.

It is possible that the problem is caused by incorrect HTTP response code. If post-back notification was successfully processed validate.php must return HTTP code 200. If your website replied with 404, 500 or any other status code except 200, post-back notification will be sent again. Some websites use additional modules\proxy servers\firewalls to change response codes.

Can you please also attach Order Messages for a new order with the same problem? Please change one line in https://github.com/ICEPAYdev/Prestashop/blob/master/validate.php#L72 file before creating the order. There is an incompatibility problem in the current module version and previous order state is not shown in order message log. We will release a new payment module version soon designed specially for Prestashop 1.6.X.

`$msg->message = "{$icepay->getTransactionString()} | Old state: {$order->order_state}; ";`
Change to 
`$msg->message = "{$icepay->getTransactionString()} | Old state: {$this->byPrestaStatus[$order->getCurrentOrderState()->id]}; ";`

![capture](https://cloud.githubusercontent.com/assets/15112699/15198243/e4af6ca8-17de-11e6-889e-492e8a60e8f2.PNG)

**Comment by tsygankov** [comment]
Hi,

We've checked your payment transaction server logs. When ICEPAY payment gateway sent a first post-back notification HTTP response code was 500. Most likely the post-back notification was processed by your website, however for some an internal server error occurred right after the transaction information was saved to Prestashop database. ICEPAY gateway is trying to resend post-back notifications several times if previous post-back notification failed (HTTP response code was not 200). Please check your website error logs for the source of the internal server error when ICEPAY payment gateway is calling validate.php

Please let me know if you need any additional information.

**Comment by luwes** [comment]
Thanks for the very useful information @tsygankov. I will try to search for the 500 error.
This is the last double payment we received (there was one payment which didn't have the issue as well)

![screen shot 2016-05-17 at 09 32 25](https://cloud.githubusercontent.com/assets/360826/15314492/7ac44254-1c12-11e6-9ea0-561670c38134.png)

**Comment by luwes** [comment]
We have this in the access logs, the first time the double payment occurred.
`ip.of.user.removed - - [08/May/2016:14:18:54 +0200] "POST /modules/icepay/validate.php HTTP/1.1" 500 - "http://icepay.eu/#API" "icepay.eu"`

**Comment by tsygankov** [comment]
Can you please also check PHP error log? It is possible that there was a problem when a new transaction status was saved to database. If you are using MySQL > 5.5.3 please try enabling innodb_print_all_deadlocks to save information about all deadlocks in InnoDB user transactions to the mysqld error log.
Please let me know if you are using MySQL replication.

**Comment by Ady78** [comment]
Guys, I have the same issue and I use Cash on delivery (COD).
It seems that all duplicate payments appear for each change of the order status, where that order status has checked the invoice option in Orders->Statuses.
Please help with a solution.
Thx

**Comment by Ady78** [comment]
Hi again,
I just discovered that it is about setting "Set the order as paid" in the statuses from Orders->Statuses. All duplicate payments appear for each change of the order status, where that order status has checked the "Set the order as paid" option in Orders->Statuses.
Leaving check the "Set the order as paid" for the last status solves the issue.
Regards,
Adrian

**Comment by diynevala** [comment]
Hi guys, I had to register only to reply to this thread. I am using Prestashop 1.6.0.14, MySQL 5.6, and PHP 5.5. I found out that every time I set status to "delivered", a new invoice gets created, and along with that a new payment is created. After a comparison with latest version of prestashop, I found the problem and a fix that was already made in the later version: 

https://github.com/PrestaShop/PrestaShop/commit/0f4a359338df39a2471bb46f62fe11aa30177f3c

I guess the SQL was faulty, but it had worked on an earlier version of the mysql?


I hope this helps.
